yjit/src/backend/ir.rs

```diff
@@ -899,21 +917,30 @@ pub struct Assembler
 
     /// Context for generating the current insn
     pub ctx: Context,
+
+    /// Parallel vec with insns
+    /// Past Context for this insn
+    pub(super) insn_ctx: Vec<Context>,
```

Comment:
Hmmmm. Is this storing a context for each YARV insn or each backend IR insn?

Reply:
The latter. At the time of `asm.compile`, the latest `Context` isn't useful for anything. So if we want to do something using `Context` in `Assembler`, e.g. identifying a stack slot is an immediate, this is needed regardless of side exits.

Another way I thought of was to introduce a backend IR that's inserted every time `Context` is detected to be changed on `push_insn`, which would limit the number of `Context`s on memory during compilation time. That would add another backend IR, so I didn't go for it, but the logic to propagate `Context` from an `Assembler` to another would be simpler than the current implementation.

Did you have other ideas as to how we're going to use `Context` from `Assembler`?

Reply:
I stopped cloning `Context` for each insn at https://github.com/ruby/ruby/pull/7712/commits/f322d2ab24bcb179bcfcb60dbaca0ca788bf120a.

The performance impact became much smaller.

```
before: ruby 3.3.0dev (2023-04-14T18:21:29Z master d83e59e6b8) +YJIT [x86_64-linux]
after: ruby 3.3.0dev (2023-04-14T19:17:44Z yjit-side-exit f322d2ab24) +YJIT [x86_64-linux]

----------  -----------  ----------  ----------  ----------  ------------  -------------
bench       before (ms)  stddev (%)  after (ms)  stddev (%)  before/after  after 1st itr
30k_ifelse  1482.5       0.0         1541.2      0.0         0.96          0.96
----------  -----------  ----------  ----------  ----------  ------------  -------------
```